File: q1-25/q23.py
# 2020.07.26
# May not have time to do it on 27th, so...
# Problem Statement
# https://leetcode.com/problems/merge-k-sorted-lists/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# one by one compare
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        
        # get rid of empty linked lists
        lists = [i for i in lists if i]
        
        # check empty case
        if len(lists) == 0: return None

        # initialize pointers
        list_ptr = []
        for list_head in lists:
            list_ptr.append(list_head)            
        
        answer_ptr = ListNode()
        answer_head = ListNode()
        first_time = True

        # loop until finish going through all the nodes
        while len(list_ptr) != 0:
            
            # pick the smallest value and add
            smallest = inf
            smallest_idx = 0
            for i in range(0, len(list_ptr)):
                if list_ptr[i] is not None and list_ptr[i].val < smallest:
                    smallest = list_ptr[i].val
                    smallest_idx = i
            answer_ptr.val = smallest

            if first_time:
                answer_head = answer_ptr
                first_time = False
            
            # move the ptr which points to the smallest value to the next
            if list_ptr[smallest_idx].next is not None:
                list_ptr[smallest_idx] = list_ptr[smallest_idx].next
            # if has been traversed, remove from the list
            else:
                list_ptr.remove(list_ptr[smallest_idx])
                
            # if all have been traversed
            if len(list_ptr) == 0:
                return answer_head
            else:
                # should add other nodes
                answer_ptr.next = ListNode()
                answer_ptr = answer_ptr.next
                
                # only one list left, can do early return
                if len(list_ptr) == 1:
                    answer_ptr.val = list_ptr[0].val
                    answer_ptr.next = list_ptr[0].next
                    return answer_head
                        
        return answer_head

File: Amazon/q23.java
// 2022.05.17
// Problem Statement:
// https://leetcode.com/problems/merge-k-sorted-lists/

// idea: direct merging K lists, each time find the min among all heads, attach the answer's tail to the min,
// detach the min from the linked list, and move its head forward
// can check this faster recursive solution (based on merging 2 sorted lists):
// https://leetcode.com/problems/merge-k-sorted-lists/discuss/10522/My-simple-java-Solution-use-recursion
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode curr_min = null;
    public int curr_min_list_idx;
    public ListNode head = null;
    public ListNode tail = null;
    public boolean first_time;
    public boolean done;
    public int count = 0;
    
    public void mergeKListsHelper(ListNode[] lists) {
        first_time = true;
        done = true;
        for (int i=0; i<lists.length; i++) {
            if (lists[i]==null) {
            } else {
                done = false;
                if (first_time) {
                    curr_min = lists[i];
                    curr_min_list_idx = i;
                    first_time = false;
                } else if (lists[i].val<curr_min.val) {
                    curr_min = lists[i];
                    curr_min_list_idx = i;
                } 
            }
        }
        if (done) return;
        lists[curr_min_list_idx] = lists[curr_min_list_idx].next;
        if (head==null && tail==null) {
            head = curr_min;
            tail = curr_min;
        }
        tail.next = curr_min;
        curr_min.next = null;
        tail = curr_min;
        count++;
    }
    
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length==0 || (lists.length==1 && lists[0]==null)) return null;
        while (!done) {
            mergeKListsHelper(lists);
        }
        return head;
    }
}